### ISSUE TITLE
What is proper syntax of exclude_paths?

### ISSUE THREAD

**Issue by remik** [codeclimate/codeclimate#466]
I tried few versions of exclude_paths but it is still not excluding.

My command:
```
[DEBUG] docker run: ["docker", "run", "--name", "cc-engines-fixme-stable-12988981-7e87-4676-82d5-059c5412e541", "--cap-drop", "all", "--label", "com.codeclimate.label=a1a27ef8-ac7f-4a74-b184-2f41e6aba20a", "--memory", "512000000", "--memory-swap", "-1", "--net", "none", "--rm", "--volume", "/mnt1/home/jenkins/workspace/web-codeclimate:/code:ro", "--volume", "/tmp/cc/e8a1336d-40f4-41d3-82e4-c872ae0dd802:/config.json:ro", "--user", "9000:9000", "codeclimate/codeclimate-fixme"]
```

My config:
```
---
engines:
  duplication:
    enabled: false
    config:
      languages:
      - javascript
      - python
  eslint:
    enabled: false
  fixme:
    enabled: true
  radon:
    enabled: true
exclude_paths:
- "static/vendor/**/*"
- "scripts/chatscript/**/*"
- "node_modules/**/*"
```

Example:
```
[DEBUG] fixme:stable engine output: {"categories":["Bug Risk"],"check_name":"TODO","description":"TODO found","location":{"lines":{"begin":170,"end":170},"path":"./node_modules/karma-jasmine/lib/adapter.js"},"type":"issue"}
```

**Comment by pbrisbin** [comment]
Hi @remik, sorry for the trouble here. The exclude paths look correct to me, `- node_modules/` is also another option but either the pattern or directory name should work.

I just verified that things are working here:

I have a project with one fixme issue:

```
% CODECLIMATE_DEBUG=1 codeclimate analyze -e fixme
Starting analysis
[DEBUG] fixme:stable engine config: {"enabled":true,"include_paths":["tmp/","lib/","app/",".env.development.example","log/","docker/",".ruby-version",".env.test","Gemfile.lock","Dockerfile",".codeclimate.yml",".dockerignore","config.ru","Makefile",".env",".gitignore","circle.yml","Gemfile",".rubocop.yml","LICENSE","tags","Rakefile"]}
[DEBUG] docker run: ["docker", "run", "--name", "cc-engines-fixme-stable-29d5dd95-5ba4-4d94-9342-09ed332c2d0c", "--cap-drop", "all", "--label", "com.codeclimate.label=35c23b51-e25f-4166-ac96-bc090ed04f5a", "--memory", "512000000", "--memory-swap", "-1", "--net", "none", "--rm", "--volume", "/home/patrick/code/codeclimate/api:/code:ro", "--volume", "/tmp/cc/63c21ac8-f29a-4ee2-b0f3-4ede6d443b00:/config.json:ro", "--user", "9000:9000", "codeclimate/codeclimate-fixme"]
[DEBUG] fixme:stable engine output: {"categories":["Bug Risk"],"check_name":"TODO","description":"TODO found","location":{"lines":{"begin":7,"end":7},"path":"app/controllers/v1/user/repo_permissions_controller.rb"},"type":"issue"}
[DEBUG] fixme:stable engine stderr: 

== app/controllers/v1/user/repo_permissions_controller.rb (1 issue) ==
7: TODO found [fixme]

Analysis complete! Found 1 issue.
```

And if I exclude by a similar pattern, it goes away:

```
% echo '- app/controllers/**/*' >> .codeclimate.yml
% tail -6 .codeclimate.yml
exclude_paths:
- "*.md"
- bin/
- config/
- spec/
- app/controllers/**/*
% CODECLIMATE_DEBUG=1 codeclimate analyze -e fixme
Starting analysis
[DEBUG] fixme:stable engine config: {"enabled":true,"include_paths":["tmp/","lib/","app/models/","app/services/","app/serializers/",".env.development.example","log/","docker/",".ruby-version",".env.test","Gemfile.lock","Dockerfile",".codeclimate.yml",".dockerignore","config.ru","Makefile",".env",".gitignore","circle.yml","Gemfile",".rubocop.yml","LICENSE","tags","Rakefile"]}
[DEBUG] docker run: ["docker", "run", "--name", "cc-engines-fixme-stable-a0db3a45-f17f-4125-8716-e20f55e0ea91", "--cap-drop", "all", "--label", "com.codeclimate.label=b5a79a75-45af-4f57-993d-b6c78b02f63c", "--memory", "512000000", "--memory-swap", "-1", "--net", "none", "--rm", "--volume", "/home/patrick/code/codeclimate/api:/code:ro", "--volume", "/tmp/cc/719125a6-8005-4e02-989e-3e4eb0a7ee2e:/config.json:ro", "--user", "9000:9000", "codeclimate/codeclimate-fixme"]
[DEBUG] fixme:stable engine stderr: 

Analysis complete! Found 0 issues.                         
```

I notice that in my case the issues are reported as `path: foo/bar`, but yours are `path: ./foo/bar` -- this could be why the exclude path isn't matching for you, and may be because you're using an out of date image that has an old bug. Can you confirm you're on the latest of everything and see if the issue persists?

```
docker pull codeclimate/codeclimate
codeclimate engines:install
codeclimate analyze -e fixme
```

Let me know.

**Comment by remik** [comment]
I installed it yesterday so I assume it is latest.
I run your commnads:
```
$ docker pull codeclimate/codeclimate
Using default tag: latest
latest: Pulling from codeclimate/codeclimate
2f3f3e5e133b: Already exists
2654c654a6e7: Already exists
412e64056adf: Already exists
a3ed95caeb02: Already exists
358c606c9e8c: Already exists
046233a2b112: Already exists
7afcb31331eb: Already exists
c3b0011af4e4: Already exists
b21ee26695b6: Already exists
06e3f9018a4d: Already exists
8bc7aaf2800a: Already exists
a668d0cdafcd: Already exists
Digest: sha256:9b387b747ac406d66e86cacac7336696e57635e4b41cc3d51b99d94a48d308c0
Status: Image is up to date for codeclimate/codeclimate:latest
$ docker run   --interactive --tty --rm   --env CODECLIMATE_CODE="$PWD"   --volume "$PWD":/code   --volume /var/run/docker.sock:/var/run/docker.sock   --volume /tmp/cc:/tmp/cc   codeclimate/codeclimate engines:install
Pulling docker images.
latest: Pulling from codeclimate/codeclimate-fixme
Digest: sha256:7a97f603f9ae200dc3604fe5531a11c33d3d5220fae03fb2ca79f532cc497aed
Status: Image is up to date for codeclimate/codeclimate-fixme:latest
latest: Pulling from codeclimate/codeclimate-radon
efd26ecc9548: Already exists
a3ed95caeb02: Already exists
364b21a1aba7: Already exists
14de0ac9af46: Already exists
f06abae204f7: Already exists
0146945d3ce0: Already exists
01284cc4d3b7: Already exists
3ca254603383: Already exists
086952cfbea1: Already exists
523aa3ae2b5a: Already exists
77feb445794f: Already exists
c99992aa6c39: Already exists
e88bb41a2fd8: Already exists
Digest: sha256:7e4334b768dcae8bedd53573793773e08b79868423731e748df4ea5c9d589617
Status: Image is up to date for codeclimate/codeclimate-radon:latest
$ docker run   --interactive --tty --rm   --env CODECLIMATE_CODE="$PWD" --env CODECLIMATE_DEBUG=1   --volume "$PWD":/code   --volume /var/run/docker.sock:/var/run/docker.sock   --volume /tmp/cc:/tmp/cc   codeclimate/codeclimate analyze -e fixme
....
[DEBUG] fixme:stable engine output: {"categories":["Bug Risk"],"check_name":"TODO","description":"TODO found","location":{"lines":{"begin":116,"end":116},"path":"./node_modules/karma-coverage/node_modules/ibrik/node_modules/fileset/node_modules/glob/common.js"},"type":"issue"}
```

I also found:
```
[DEBUG] Couldn't include because part of path doesn't exist. path="./."
[DEBUG] fixme:stable engine config: {"enabled":true,"include_paths":["./"]}
```

Maybe something is wrong with `--volumes` params? I copied it from README.

**Comment by pbrisbin** [comment]
Oh interesting, I didn't realize you were doing a `docker run`-based invocation. I did re-test locally using you exact invocation and it worked as expected. Can show the output of `echo $PWD`?

Ping @wfleming I know you worked on the code in the area that emits this message:

```
[DEBUG] Couldn't include because part of path doesn't exist. path="./."
```

Maybe you have some ideas?

**Comment by remik** [comment]
Here you go:
```
$ echo $PWD
/home/remik/workspace/web-codeclimate
```
I forgot to install wrapper. I will install it check again.

**Comment by remik** [comment]
Found issue!
You are wrong handling relative paths. 
I put `.` in `<path>` param so it is current dir but it breaks exclude_paths.

Should I keep this ticket to fix it or you have internal process?

**Comment by remik** [comment]
I think my above command is wrong - I will edit it.

**Comment by pbrisbin** [comment]
Hi @remik are you still running into issues here? If so, can you restate what's going on -- the thread's a little hard to follow. If not, I'll go ahead and close. Thanks.

**Comment by remik** [comment]
I changed to absolute path and it is fine.
